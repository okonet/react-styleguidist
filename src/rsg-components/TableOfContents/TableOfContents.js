import React, { Component, PropTypes } from 'react';
import { filterComponentsByName, getFilterRegExp } from '../../utils/utils';
import ComponentsList from 'rsg-components/ComponentsList';
import TableOfContentsRenderer from 'rsg-components/TableOfContents/TableOfContentsRenderer';

export default class TableOfContents extends Component {
	static propTypes = {
		sections: PropTypes.array.isRequired,
	};

	constructor(props) {
		super(props);

		this.state = {
			searchTerm: '',
		};
	}

	getComponents(components, searchTerm) {
		return filterComponentsByName(components || [], searchTerm);
	}

	getSections(sections = [], searchTerm) {
		const regExp = getFilterRegExp(searchTerm);
		return sections.reduce((filteredSections, { name, components: subComponents = [], sections: subSections }) => {
			subComponents = this.getComponents(subComponents, searchTerm);
			if (subComponents.length || !searchTerm || regExp.test(name)) {
				filteredSections.push({
					heading: true,
					name,
					content: this.renderLevel(subComponents, subSections, searchTerm),
				});
			}
			return filteredSections;
		}, []);
	}

	renderLevel(components, sections, searchTerm) {
		const items = [
			...this.getComponents(components, searchTerm),
			...this.getSections(sections, searchTerm),
		];
		return (
			<ComponentsList items={items} />
		);
	}

	render() {
		const { searchTerm } = this.state;
		const { sections } = this.props;

		// If there is only one section, we treat it as a root section
		// In this case the name of the section won't be rendered and it won't get left padding
		const content = sections.length === 1
			? this.renderLevel(sections[0].components, sections[0].sections, searchTerm)
			: this.renderLevel(null, sections, searchTerm);
		return (
			<TableOfContentsRenderer
				searchTerm={searchTerm}
				onSearchTermChange={searchTerm => this.setState({ searchTerm })}
			>
				{content}
			</TableOfContentsRenderer>
		);
	}
}
