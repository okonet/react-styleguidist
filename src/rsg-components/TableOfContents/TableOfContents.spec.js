import React from 'react';
import noop from 'lodash/noop';
import TableOfContents from './TableOfContents';
import { TableOfContentsRenderer } from './TableOfContentsRenderer';

const components = [
	{
		name: 'Button',
	},
	{
		name: 'Input',
	},
	{
		name: 'Textarea',
	},
];

const sections = [
	{
		name: 'Introduction',
		content: 'intro.md',
	},
	{
		name: 'Buttons',
		components: [
			{
				name: 'Button',
			},
		],
	},
	{
		name: 'Forms',
		components: [
			{
				name: 'Input',
			},
			{
				name: 'Textarea',
			},
		],
	},
];

it('should render a renderer', () => {
	const actual = shallow(
		<TableOfContents sections={[{ components }]} />
	);

	expect(actual).toMatchSnapshot();
});

it('should render renderer with sections with nested components', () => {
	const actual = shallow(
		<TableOfContents sections={sections} />
	);

	expect(actual).toMatchSnapshot();
});


it('should filter list when search field contains a query', () => {
	const searchTerm = 'but';
	const actual = shallow(
		<TableOfContents sections={[{ components }]} />
	);

	expect(actual).toMatchSnapshot();

	actual.setState({ searchTerm });

	expect(actual).toMatchSnapshot();
});

it('should render a filtered list, should hide empty sections', () => {
	const searchTerm = 'inp';
	const actual = shallow(
		<TableOfContents sections={sections} />
	);

	expect(actual).toMatchSnapshot();

	actual.setState({ searchTerm });

	expect(actual).toMatchSnapshot();
});

it('should filter section names', () => {
	const searchTerm = 'frm';
	const actual = shallow(
		<TableOfContents sections={sections} />
	);

	expect(actual).toMatchSnapshot();

	actual.setState({ searchTerm });

	expect(actual).toMatchSnapshot();
});

it('renderer should render table of contents', () => {
	const searchTerm = 'foo';
	const actual = shallow(
		<TableOfContentsRenderer
			classes={{}}
			items={<div>foo</div>}
			searchTerm={searchTerm}
			onSearchTermChange={noop}
		/>
	);

	expect(actual).toMatchSnapshot();
});

it('should call a callback when input value changed', () => {
	const onSearchTermChange = jest.fn();
	const searchTerm = 'foo';
	const newSearchTerm = 'bar';
	const actual = shallow(
		<TableOfContentsRenderer
			classes={{}}
			items={<div>foo</div>}
			searchTerm={searchTerm}
			onSearchTermChange={onSearchTermChange}
		/>
	);

	actual.find('input').simulate('change', {
		target: {
			value: newSearchTerm,
		},
	});

	expect(onSearchTermChange).toBeCalledWith(newSearchTerm);
});
