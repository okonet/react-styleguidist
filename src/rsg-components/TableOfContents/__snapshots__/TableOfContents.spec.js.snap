exports[`test renderer should render table of contents 1`] = `
<div>
  <div>
    <div>
      <input
        onChange={[Function]}
        placeholder="Filter by name"
        value="foo" />
    </div>
  </div>
</div>
`;

exports[`test should filter list when search field contains a query 1`] = `
<_class
  onSearchTermChange={[Function]}
  searchTerm="">
  <_class
    items={
      Array [
        Object {
          "name": "Button",
        },
        Object {
          "name": "Input",
        },
        Object {
          "name": "Textarea",
        },
      ]
    } />
</_class>
`;

exports[`test should filter list when search field contains a query 2`] = `
<_class
  onSearchTermChange={[Function]}
  searchTerm="but">
  <_class
    items={
      Array [
        Object {
          "name": "Button",
        },
      ]
    } />
</_class>
`;

exports[`test should filter section names 1`] = `
<_class
  onSearchTermChange={[Function]}
  searchTerm="">
  <_class
    items={
      Array [
        Object {
          "content": <_class
            items={Array []} />,
          "heading": true,
          "name": "Introduction",
        },
        Object {
          "content": <_class
            items={
                  Array [
                        Object {
                          "name": "Button",
                        },
                      ]
            } />,
          "heading": true,
          "name": "Buttons",
        },
        Object {
          "content": <_class
            items={
                  Array [
                        Object {
                          "name": "Input",
                        },
                        Object {
                          "name": "Textarea",
                        },
                      ]
            } />,
          "heading": true,
          "name": "Forms",
        },
      ]
    } />
</_class>
`;

exports[`test should filter section names 2`] = `
<_class
  onSearchTermChange={[Function]}
  searchTerm="frm">
  <_class
    items={
      Array [
        Object {
          "content": <_class
            items={Array []} />,
          "heading": true,
          "name": "Forms",
        },
      ]
    } />
</_class>
`;

exports[`test should render a filtered list, should hide empty sections 1`] = `
<_class
  onSearchTermChange={[Function]}
  searchTerm="">
  <_class
    items={
      Array [
        Object {
          "content": <_class
            items={Array []} />,
          "heading": true,
          "name": "Introduction",
        },
        Object {
          "content": <_class
            items={
                  Array [
                        Object {
                          "name": "Button",
                        },
                      ]
            } />,
          "heading": true,
          "name": "Buttons",
        },
        Object {
          "content": <_class
            items={
                  Array [
                        Object {
                          "name": "Input",
                        },
                        Object {
                          "name": "Textarea",
                        },
                      ]
            } />,
          "heading": true,
          "name": "Forms",
        },
      ]
    } />
</_class>
`;

exports[`test should render a filtered list, should hide empty sections 2`] = `
<_class
  onSearchTermChange={[Function]}
  searchTerm="inp">
  <_class
    items={
      Array [
        Object {
          "content": <_class
            items={
                  Array [
                        Object {
                          "name": "Input",
                        },
                      ]
            } />,
          "heading": true,
          "name": "Forms",
        },
      ]
    } />
</_class>
`;

exports[`test should render a renderer 1`] = `
<_class
  onSearchTermChange={[Function]}
  searchTerm="">
  <_class
    items={
      Array [
        Object {
          "name": "Button",
        },
        Object {
          "name": "Input",
        },
        Object {
          "name": "Textarea",
        },
      ]
    } />
</_class>
`;

exports[`test should render renderer with sections with nested components 1`] = `
<_class
  onSearchTermChange={[Function]}
  searchTerm="">
  <_class
    items={
      Array [
        Object {
          "content": <_class
            items={Array []} />,
          "heading": true,
          "name": "Introduction",
        },
        Object {
          "content": <_class
            items={
                  Array [
                        Object {
                          "name": "Button",
                        },
                      ]
            } />,
          "heading": true,
          "name": "Buttons",
        },
        Object {
          "content": <_class
            items={
                  Array [
                        Object {
                          "name": "Input",
                        },
                        Object {
                          "name": "Textarea",
                        },
                      ]
            } />,
          "heading": true,
          "name": "Forms",
        },
      ]
    } />
</_class>
`;
